gradle/invoke-javadoc.gradle

```diff
@@ -0,0 +1,335 @@
+        }
+        // append prettify to scripts
+        ant.concat(destfile: "${javadocDestDir}/script.js", append: "true", fixlastline: "true", encoding: "UTF-8") {
+          filelist(dir: project(':lucene').file("tools/prettify"), files: "prettify.js inject-javadocs.js")
+        }
+        ant.fixcrlf(srcdir: javadocDestDir, includes: "stylesheet.css script.js", eol: "lf", fixlast: "true", encoding: "UTF-8")
+      }
+    }
+  }
+}
+
+configure(subprojects.findAll { it.path.startsWith(':lucene') && it.path != ':lucene:core' }) {
+
+  plugins.withType(JavaPlugin) {
+    invokeJavadoc {
+      dependsOn ':lucene:core:invokeJavadoc'
+
+      inputs.property('linkHref', inputs.properties.linkHref + [ '../core' ])
+
+      doLast {
+        // fix for Java 11 Javadoc tool that cannot handle split packages between modules correctly (by removing all the packages which are part of lucene-core)
+        // problem description: [https://issues.apache.org/jira/browse/LUCENE-8738?focusedCommentId=16818106&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16818106]
+        ant.local(name: "element-list-regex") // contains a regex for all package names which are in lucene-core's javadoc
+        ant.loadfile(property: "element-list-regex", srcFile: "${project.javadocRoot}/core/element-list", encoding: "utf-8") {
+          filterchain {
+            tokenfilter(delimoutput: "|") {
+              replacestring(from: ".", to: "\\.")
+            }
+          }
+        }
+        ant.replaceregexp(
+            encoding: "UTF-8",
+            file: "${project.javadocDestDir}/element-list",
+            byline: "true",
+            match: "^(\${element-list-regex})\$",
+            replace: "")
+
+        //println("dest dir = ${project.javadocDestDir}")
+        //println("linksource = ${inputs.properties.linksource}")
+        //println("link href = ${inputs.properties.linkHref}")
+      }
+    }
+  }
+}
+
+configure(subprojects.findAll { it.path.startsWith(':lucene:analysis') }) {
+  ext {
+    javadocDestDir = "${javadocRoot}/analyzers-${project.name}"
+  }
+
+  plugins.withType(JavaPlugin) {
+    invokeJavadoc {
+      if (project.path != ':lucene:analysis:common') {
+        dependsOn ':lucene:analysis:common:invokeJavadoc'
+        inputs.property('linkHref', inputs.properties.linkHref + [ '../analyzers-common' ])
+      }
+    }
+  }
+}
+
+configure(project(':lucene:benchmark')) {
+  plugins.withType(JavaPlugin) {
+    invokeJavadoc {
+      [':lucene:memory',
+       ':lucene:highlighter',
+       ':lucene:analysis:common',
+       ':lucene:queryparser',
+       ':lucene:facet',
+       ':lucene:spatial-extras'].collect { path ->
+        dependsOn "${path}:invokeJavadoc"
+      }
+      inputs.property('linkHref', inputs.properties.linkHref + [ '../memory', '../highlighter', '../analyzers-common', '../queryparser', '../facet', '../spatial-extras' ])
```

Comment:
These relative paths should ideally go through project names and transform into relative paths. So that in case we change project paths it'd fail early and not silently point at a wrong folder?

[":lucene:memory", ":lucene:highlighter", ...].collect { prjPath -> "../" + project(prjPath).name }

Reply:
The relative paths are not necessarily inferred from their project name.
- ":lucene:analysis:icu" => docs/analyzers-icu
- ":solr:solrj" => docs/solr-solrj
- ":solr:contrib:extraction" => docs/solr-cell (!)
...

How can we properly map the projects and javadoc output directories (without messing up the task definitions any further)...?

Reply:
This code is emulating ant - hence the folder name difference (I created gradle project structure with focus on clarity, not conformance to ant). You could set up a map of project paths to those "ant" folder names and then look up the corresponding path from project but I'm fine if you don't want to spend any more time on it.

Reply:
I'd like to leave it as it is for now, because it's still not clear to me where/how the map should be placed.
One thing I noticed is, we might also need this map (project to javadoc folder name - in other words, ant project name) when publishing the docs to apache.org web site, not just for resolving internal links?
e.g.)
https://lucene.apache.org/core/8_4_0/analyzers-icu/index.html
https://lucene.apache.org/solr/8_4_1/solr-solrj/index.html

Reply:
I don't think we should try to emulate ant here, Tomoko. Ideally, I'd rather have the documentation follow gradle project path here. So ".../core/[ver]/analysis/icu/index.html", for example. The fewer exceptions and hand-written mappings the easier it is to change things later on.

Reply:
> Ideally, I'd rather have the documentation follow gradle project path here. So ".../core/[ver]/analysis/icu/index.html", for example. 

I agree with that it makes things easy and i myself would love to throw away the convention come from Ant for simplicity. 
But changing folder structure on lucene.apahce.org would be a bit big change for devs, maybe it's required to gain an agreement with others...?

Reply:
I think this change should follow dropping ant build entirely from master. Then things become simpler to manage. These links are prefixed with version number anyway so I don't think there is harm in the docs structure changing for /9_0_0/....

Reply:
Ok I will create another patch to make the folder structure follow the gradle project path, instead of ant one.

Reply:
You can also create a placeholder issue for such a change, Tomoko. I don't think we can introduce it until we drop ant anyway (?).